### PR TITLE
Fallback architecture

### PR DESCRIPTION
**Pull request by SimonRichardson**
We should always have an architecture, even if we get "all" we should
fallback to either the model constraints or the default architecture.

That should ensure that we're never passing an empty string to the API.

## QA steps

We can use the microk8s-testbed charm here as it exhibits the s390x as the
default architecture in the info API.

----

This pretty much requires us to test all permutations:

```sh
$ juju bootstrap lxd test
$ juju add-model one --config charm-hub-url="https://api.staging.snapcraft.io"
```

### Without any constraints

```sh
$ juju deploy microk8s-testbed
```

### With constraints

```sh
$ juju deploy microk8s-testbed --constraints="arch=amd64"
```

### Model constraints

```sh
$ juju bootstrap --bootstrap-constraints arch=amd64 lxd test2
$ juju add-model two --config charm-hub-url="https://api.staging.snapcraft.io"
$ juju deploy microk8s-testbed
```

### Charm Architecture in bundles

The following deploys fine one aws:

```sh
series: focal
applications:
  ubuntu-s390x:
    charm: ch:ubuntu
    num_units: 1
    constraints: arch=s390x
```

The following should fail:

```sh
series: focal
applications:
  ubuntux:
    charm: ch:ubuntu
    num_units: 1
    constraints: arch=amd64
  ubuntu-s390x:
    charm: ch:ubuntu
    num_units: 1
    constraints: arch=s390x
```

```
➜ juju deploy ./bundle.yaml
Located charm "ubuntu" in charm-hub
Located charm "ubuntu" in charm-hub
Executing changes:
- upload charm ubuntu from charm-hub for series focal with architecture=s390x
- deploy application ubuntu-s390x from charm-hub on focal using ubuntu
- deploy application ubuntux from charm-hub on focal using ubuntu
ERROR cannot deploy bundle: unexpected application architecture (amd64), charm already exists with architecture (s390x)
```

## Bug reference

https://bugs.launchpad.net/snapstore-server/+bug/1908427

### CODE REVIEW

**Comment by hmlanigan** [comment]
Found in qa

$ juju deploy ubuntu-juju-qa --constraints="arch=arm64"
Located charm "ubuntu-juju-qa" in charm-hub, revision 1
Deploying "ubuntu-juju-qa" from charm-hub charm "ubuntu-juju-qa", revision 1 in channel latest/stable

leads too:
		"platform" : {
			"architecture" : "s390x",
			"os" : "ubuntu",
			"series" : "focal"
		}

**Review by hmlanigan — COMMENTED** [review]
qa good, will finish the code review part in the AM.

**Comment by hmlanigan** [comment]
QA: charmstorm charm origin does not reflect model constraint architecture, though machine deployment fails as expected with no matching agent binaries

$ juju add-model default
Added 'default' model on aws/us-east-1 with credential 'credentials' for user 'admin'
$ juju set-model-constraints "arch=s390x"
$ juju get-model-constraints
arch=s390x
$ juju deploy ubuntu
Located charm "ubuntu" in charm-hub, revision 17
Deploying "ubuntu" from charm-hub charm "ubuntu", revision 17 in channel latest/stable
$ juju deploy cs:ubuntu csubuntu
Located charm "ubuntu" in charm-store, revision 16
Deploying "csubuntu" from charm-store charm "ubuntu", revision 16 in channel stable

juju:PRIMARY> db.applications.find({},{ "charmurl" :1,"charm-origin" :1}).pretty()
{
	"_id" : "30ba7b60-f6ea-4699-859a-b73e8b8bf5f9:ubuntu",
	"charmurl" : "ch:ubuntu-17",
	"charm-origin" : {
		"source" : "charm-hub",
...
		"platform" : {
			"architecture" : "s390x",
			"os" : "ubuntu",
			"series" : "focal"
		}
	}
}
{
	"_id" : "30ba7b60-f6ea-4699-859a-b73e8b8bf5f9:csubuntu",
	"charmurl" : "cs:ubuntu-16",
	"charm-origin" : {
		"source" : "charm-store",
...
		"platform" : {
			"architecture" : "amd64",    <-- expected s390x
			"os" : "ubuntu",
			"series" : "focal"
		}
	}
}

**Comment by hmlanigan** [comment]
QA: arch constraints defined in bundle not saved to origin platform correctly.

bundle and juju status: https://pastebin.canonical.com/p/RGRBk27gyY/. status as expected.

unexpected db results: https://pastebin.canonical.com/p/hsQ8GjvHNf/

**Review by hmlanigan — COMMENTED** [review]
mostly nits, though a few qa pieces to be addressed.

**Comment by hmlanigan** [comment]
Not having an architecture for a CS charm, or any, causes model migration to fail.  `migrating: aborted, removing model from target controller: model data transfer failed, model export failed: arch "" not valid`

CS charms deployed by bundle, have no architecture, though CH ones do.  The jenkins migration tests do not fail, as the charms are deployed individually, not by bundle.

File: apiserver/facades/client/charms/client.go [hunk 1]
```diff
@@ -570,6 +594,18 @@ func (a *API) resolveOneCharm(arg params.ResolveCharmWithChannel, mac *macaroon.
 	return result
 }
 
+func validateOrigin(origin params.CharmOrigin, schema string) error {
+	if (corecharm.Local.Matches(origin.Source) && !charm.Local.Matches(schema)) ||
+		(corecharm.CharmStore.Matches(origin.Source) && !charm.CharmStore.Matches(schema)) ||
+		(corecharm.CharmHub.Matches(origin.Source) && !charm.CharmHub.Matches(schema)) {
+		return errors.NotValidf("origin source %q with schema", origin.Source)
+	}
+	if corecharm.CharmHub.Matches(origin.Source) && origin.Architecture == "" {
```
Comment (hmlanigan) [inline apiserver/facades/client/charms/client.go:603]: I thought, due to the migration bug https://bugs.launchpad.net/juju/+bug/1911800, architecture would be required with a CS charm too.

**Review by hmlanigan — APPROVED** [review]
LGTM, just the 1 question

File: apiserver/facades/client/charms/client.go (hunk 1 above)
Comment (SimonRichardson) [reply apiserver/facades/client/charms/client.go:603]: Well due to that bug, I think we should undo this, but as a separate PR and tackle that holistically.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**